File: app/src/main/java/piuk/blockchain/android/ui/contacts/payments/ContactPaymentRequestNotesFragment.java
package piuk.blockchain.android.ui.contacts.payments;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.inject.Inject;

import piuk.blockchain.android.R;
import piuk.blockchain.android.data.contacts.PaymentRequestType;
import piuk.blockchain.android.databinding.FragmentContactPaymentRequestNotesBinding;
import piuk.blockchain.android.injection.Injector;
import piuk.blockchain.android.ui.base.BaseFragment;
import piuk.blockchain.android.ui.customviews.MaterialProgressDialog;
import piuk.blockchain.android.ui.customviews.ToastCustom;
import piuk.blockchain.android.util.ViewUtils;

public class ContactPaymentRequestNotesFragment extends BaseFragment<ContactPaymentRequestView, ContactsPaymentRequestPresenter>
        implements ContactPaymentRequestView {

    public static final String ARGUMENT_REQUEST_TYPE = "request_type";
    public static final String ARGUMENT_ACCOUNT_POSITION = "account_position";
    public static final String ARGUMENT_CONTACT_ID = "contact_id";
    public static final String ARGUMENT_SATOSHIS = "satoshis";

    @Inject ContactsPaymentRequestPresenter paymentRequestPresenter;
    private FragmentContactPaymentRequestNotesBinding binding;
    private MaterialProgressDialog progressDialog;
    private FragmentInteractionListener listener;

    {
        Injector.getInstance().getPresenterComponent().inject(this);
    }

    public static ContactPaymentRequestNotesFragment newInstance(PaymentRequestType requestType,
                                                                 @Nullable Integer accountPosition,
                                                                 String contactId,
                                                                 long satoshis) {
        Bundle args = new Bundle();
        args.putSerializable(ARGUMENT_REQUEST_TYPE, requestType);
        args.putString(ARGUMENT_CONTACT_ID, contactId);
        args.putLong(ARGUMENT_SATOSHIS, satoshis);
        if (accountPosition != null) {
            args.putInt(ARGUMENT_ACCOUNT_POSITION, accountPosition);
        }
        ContactPaymentRequestNotesFragment fragment = new ContactPaymentRequestNotesFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(
                inflater, R.layout.fragment_contact_payment_request_notes, container, false);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding.buttonNext.setOnClickListener(v -> {
            getPresenter().sendRequest();
            ViewUtils.hideKeyboard(getActivity());
        });

        onViewReady();
    }

    @Override
    public void contactLoaded(String name, PaymentRequestType paymentRequestType) {
        setSummary(paymentRequestType, name);
        setHint(paymentRequestType);
        binding.editTextNote.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus || !binding.editTextNote.getText().toString().isEmpty()) {
                binding.inputLayoutNote.setHint(getString(R.string.contacts_payment_request_note_hint));
            } else {
                setHint(paymentRequestType);
            }
        });
    }

    @Override
    public Bundle getFragmentBundle() {
        return getArguments();
    }

    @Override
    public String getNote() {
        return binding.editTextNote.getText().toString();
    }

    @Override
    public void finishPage() {
        if (listener != null) listener.onPageFinished();
    }

    @Override
    public void showProgressDialog() {
        progressDialog = new MaterialProgressDialog(getContext());
        progressDialog.setCancelable(false);
        progressDialog.setMessage(R.string.please_wait);
        progressDialog.show();
    }

    @Override
    public void dismissProgressDialog() {
        if (progressDialog != null) {
            progressDialog.dismiss();
            progressDialog = null;
        }
    }

    @Override
    public void showToast(@StringRes int message, @ToastCustom.ToastType String toastType) {
        ToastCustom.makeText(getContext(), getString(message), ToastCustom.LENGTH_SHORT, toastType);
    }

    @Override
    public void showSendSuccessfulDialog(String name) {
        new AlertDialog.Builder(getContext(), R.style.AlertDialogStyle)
                .setTitle(getString(R.string.contacts_payment_success_waiting_title, name))
                .setMessage(R.string.contacts_payment_success_waiting_message)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> finishPage())
                .show();
    }

    @Override
    public void showRequestSuccessfulDialog() {
        new AlertDialog.Builder(getContext(), R.style.AlertDialogStyle)
                .setTitle(R.string.contacts_payment_success_request_sent_title)
                .setMessage(R.string.contacts_payment_success_waiting_message)
                .setPositiveButton(android.R.string.ok, (dialog, which) -> finishPage())
                .show();
    }

    @Override
    protected ContactsPaymentRequestPresenter createPresenter() {
        return paymentRequestPresenter;
    }

    @Override
    protected ContactPaymentRequestView getMvpView() {
        return this;
    }

    private void setSummary(PaymentRequestType paymentRequestType, String contactName) {
        if (paymentRequestType.equals(PaymentRequestType.SEND)) {
            binding.textviewExplanation.setText(getString(R.string.contacts_payment_request_send_note, contactName));
        } else {
            binding.textviewExplanation.setText(getString(R.string.contacts_payment_request_receive_note, contactName));
        }
    }

    private void setHint(PaymentRequestType paymentRequestType) {
        if (paymentRequestType.equals(PaymentRequestType.REQUEST)) {
            binding.inputLayoutNote.setHint(getString(R.string.contacts_payment_request_receive_hint));
        } else {
            binding.inputLayoutNote.setHint(getString(R.string.contacts_payment_request_send_hint));
        }
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof FragmentInteractionListener) {
            listener = (FragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context + " must implement FragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    public interface FragmentInteractionListener {

        void onPageFinished();

    }

}

File: app/src/main/java/piuk/blockchain/android/ui/contacts/payments/ContactPaymentRequestNotesFragment.kt
package piuk.blockchain.android.ui.contacts.payments

import android.content.Context
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import kotlinx.android.synthetic.main.fragment_contact_payment_request_notes.*
import piuk.blockchain.android.R
import piuk.blockchain.android.injection.Injector
import piuk.blockchain.android.ui.account.PaymentConfirmationDetails
import piuk.blockchain.android.ui.base.BaseAuthActivity
import piuk.blockchain.android.ui.base.BaseFragment
import piuk.blockchain.android.ui.customviews.MaterialProgressDialog
import piuk.blockchain.android.ui.customviews.ToastCustom
import piuk.blockchain.android.ui.home.MainActivity
import piuk.blockchain.android.util.ViewUtils
import piuk.blockchain.android.util.extensions.getTextString
import piuk.blockchain.android.util.extensions.gone
import piuk.blockchain.android.util.extensions.inflate
import piuk.blockchain.android.util.extensions.visible
import javax.inject.Inject

class ContactPaymentRequestNotesFragment : BaseFragment<ContactPaymentRequestView, ContactsPaymentRequestPresenter>(),
        ContactPaymentRequestView {

    override val fragmentBundle: Bundle
        get() = arguments
    override val note: String
        get() = edittext_description.getTextString()

    @Inject lateinit var paymentRequestPresenter: ContactsPaymentRequestPresenter
    private var progressDialog: MaterialProgressDialog? = null
    private var listener: FragmentInteractionListener? = null

    init {
        Injector.getInstance().presenterComponent.inject(this)
    }

    override fun onCreateView(
            inflater: LayoutInflater?,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ) = container?.inflate(R.layout.fragment_contact_payment_request_notes)

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        button_send.setOnClickListener {
            presenter.sendRequest()
            ViewUtils.hideKeyboard(activity)
        }

//        setupToolbar()
        onViewReady()
    }

    override fun contactLoaded(name: String) {
        textview_to_name.text = name
        edittext_description.hint = getString(R.string.contacts_confirm_shared_with, name)
        loading_layout.gone()
        main_layout.visible()
    }

    override fun updateAccountName(name: String) {
        textview_from_address.text = name
    }

    override fun updateTotalBtc(total: String) {
        textview_total_btc.text = total
    }

    override fun updateTotalFiat(total: String) {
        textview_total_fiat.text = total
    }

    override fun finishPage() {
        listener?.onPageFinished()
    }

    override fun showProgressDialog() {
        progressDialog = MaterialProgressDialog(context)
                .apply {
                    setCancelable(false)
                    setMessage(R.string.please_wait)
                    show()
                }
    }

    override fun dismissProgressDialog() {
        progressDialog?.apply {
            dismiss()
            progressDialog = null
        }
    }

    override fun showToast(@StringRes message: Int, @ToastCustom.ToastType toastType: String) {
        ToastCustom.makeText(context, getString(message), ToastCustom.LENGTH_SHORT, toastType)
    }

    override fun showRequestSuccessfulDialog() {
        // TODO: Remove me and show success fragment instead by triggering callback
        AlertDialog.Builder(context, R.style.AlertDialogStyle)
                .setTitle(getString(R.string.contacts_payment_success_waiting_title))
                .setMessage(R.string.contacts_payment_success_waiting_message)
                .setPositiveButton(android.R.string.ok) { _, _ -> finishPage() }
                .show()
    }

    override fun createPresenter() = paymentRequestPresenter

    override fun getMvpView() = this

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is FragmentInteractionListener) {
            listener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement FragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    private fun setupToolbar() {
        // TODO: For now this isn't reset when leaving this page  
        if ((activity as AppCompatActivity).supportActionBar != null) {
            (activity as BaseAuthActivity).setupToolbar(
                    (activity as MainActivity).supportActionBar, R.string.contacts_confirm_title)
        } else {
            finishPage()
        }
    }

    interface FragmentInteractionListener {

        fun onPageFinished()

    }

    companion object {

        const val ARGUMENT_CONFIRMATION_DETAILS = "ARGUMENT_CONFIRMATION_DETAILS"
        const val ARGUMENT_CONTACT_ID = "ARGUMENT_CONTACT_ID"
        const val ARGUMENT_SATOSHIS = "ARGUMENT_SATOSHIS"

        @JvmStatic
        fun newInstance(
                confirmationDetails: PaymentConfirmationDetails,
                contactId: String,
                satoshis: Int
        ): ContactPaymentRequestNotesFragment {

            val args = Bundle().apply {
                putParcelable(ARGUMENT_CONFIRMATION_DETAILS, confirmationDetails)
                putString(ARGUMENT_CONTACT_ID, contactId)
                putInt(ARGUMENT_SATOSHIS, satoshis)
            }
            return ContactPaymentRequestNotesFragment().apply { arguments = args }
        }
    }

}
